### PR TITLE
Related #9 Cart screen

### PR DESCRIPTION
**Pull request by nikola-mladenovic**

### CODE REVIEW

File: Sunrise/Models/Money.swift [hunk 1]
```diff
@@ -9,7 +9,7 @@ struct Money: Mappable {
     // MARK: - Properties
 
     var currencyCode: String?
-    var centAmount: Int?
+    var centAmount: Double?
```
Comment (cneijenhuis) [inline Sunrise/Models/Money.swift]: Can you explain this change?

**Comment by cneijenhuis** [comment]
I think the quantity should be displayed as well (it's not editable yet though, as specified in #9).

File: Sunrise/Cells/CartItemCell.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//  CartItemCell.swift
+//  Sunrise
+//
+//  Created by Nikola Mladenovic on 5/31/16.
+//  Copyright © 2016 Commercetools. All rights reserved.
+//
+
+import UIKit
+
+class CartItemCell: UITableViewCell {
```
Comment (lauraluiz) [inline Sunrise/Cells/CartItemCell.swift]: Can we name it `CartLineItemCell` or similar? I just want to use the word `LineItem`, which is how we refer to those items. That's of course if I get right this class. By its attributes it looks like line item, although I see quantity is not there.

Comment (nikola-mladenovic) [inline Sunrise/Cells/CartItemCell.swift]: @lauraluiz yes, we can rename it to `CartLineItemCell`. This class contains properties which represent the UI element (in this case labels and an image view).

**Comment by nikola-mladenovic** [comment]
@cneijenhuis @lauraluiz - since we're not supporting increasing or decreasing quantity, how should we display it? Same as in the mock, just without + and -, or different?

File: Sunrise/Models/Money.swift (hunk 1 above)
Comment (nikola-mladenovic) [inline Sunrise/Models/Money.swift]: `return currencyFormatter.stringFromNumber(centAmount / 100) ?? ""` Cents weren't properly calculated, but now the tests are covering that as well. If we'll always have Int value for this one, we can just divide by 100.0.

Comment (cneijenhuis) [inline Sunrise/Models/Money.swift]: I would prefer to keep the `Int`, as this is what the platform will always return.

(On a sidenote, `/ 100` is correct for most western currencies, but e.g. fails for the Japanese Yen (which doesn't have cents) and others. But it's out of scope for now - couldn't find an iOS lib that deals with currencies that far with 5 min of googling).

**Comment by cneijenhuis** [comment]
Maybe it's not so obvious that it is the quantity if we don't have + and -...
Suggestion: Add it in the middle of the prices: `Price | Quantity | Total Price` WDYT?

**Comment by nikola-mladenovic** [comment]
That's reasonable for me - @lauraluiz ?

File: Sunrise/Cells/CartItemCell.swift (hunk 2 above)
Comment (lauraluiz) [inline Sunrise/Cells/CartItemCell.swift]: Ok, good to know for the future, because I just found the `LineItem` class and was getting confused :D

File: Sunrise/ViewModels/CartViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    func itemImageUrlAtIndexPath(indexPath: NSIndexPath) -> String {
+        return cart.value?.lineItems?[indexPath.row].variant?.images?.first?.url ?? ""
+    }
+
+    func itemPriceAtIndexPath(indexPath: NSIndexPath) -> String {
+        return cart.value?.lineItems?[indexPath.row].price?.value?.description ?? "N/A"
+    }
+
+    func itemTotalPriceAtIndexPath(indexPath: NSIndexPath) -> String {
```
Comment (lauraluiz) [inline Sunrise/ViewModels/CartViewModel.swift]: I would also keep here the `lineItemTotal...` name. When I saw `item` I was not sure what that was, but if named `lineItem` the association is immediate. Line item is something very specific in e-commerce, while "item" is something generic.

**Comment by lauraluiz** [comment]
With the proper title "quantity", yes, it would fit perfectly.

**Comment by nikola-mladenovic** [comment]
Ok, going down that path 👍

**Comment by lauraluiz** [comment]
👍

File: Sunrise/Cells/CartLineItemCell.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    @IBOutlet weak var priceLabel: UILabel!
+    @IBOutlet weak var quantityLabel: UILabel!
+    @IBOutlet weak var totalPriceLabel: UILabel!
+    @IBOutlet weak var skuLabel: UILabel!
+    @IBOutlet weak var sizeLabel: UILabel!
+
+    override func prepareForReuse() {
+        super.prepareForReuse()
+
+        productImageView.image = nil
```
Comment (sven-straubinger) [inline Sunrise/Cells/CartLineItemCell.swift]: From the docs it says `For performance reasons, you should only reset attributes of the cell that are not related to content, for example, alpha, editing, and selection state.`, thus we should reset the imageView in the controller only.

File: Sunrise/ViewControllers/CartViewController.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+extension CartViewController: UITableViewDataSource {
+
+    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        guard let viewModel = viewModel else { return 0 }
+        return viewModel.numberOfRowsInSection(section)
+    }
+
+    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCellWithIdentifier("CartItemCell", forIndexPath: indexPath) as! CartLineItemCell
```
Comment (sven-straubinger) [inline Sunrise/ViewControllers/CartViewController.swift]: Maybe we should rename the string `"CartItemCell"`to `"CartLineItemCell"` as well.

File: Sunrise/ViewModels/BaseViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    func alertMessageForErrors(errors: [NSError]) -> String {
+        return errors.map({
+            var alertMessage = ""
+            if let failureReason = $0.userInfo[NSLocalizedFailureReasonErrorKey] as? String {
+                alertMessage += failureReason
+            }
+            if let localizedDescription = $0.userInfo[NSLocalizedDescriptionKey] as? String {
+                alertMessage += ": \(localizedDescription)"
+            }
```
Comment (sven-straubinger) [inline Sunrise/ViewModels/BaseViewModel.swift]: Somehow it's not working when trying to access `failureReason` and `localizedDescription` via the `userInfo`-dictionary directly (I was provoking some query errors). The string was always empty (see screenshot). 

alert-1

Changing it to the code below works fine:

```
    if let failureReason = $0.localizedFailureReason {
        alertMessage += failureReason
    }

    let localizedDescription = $0.localizedDescription
    alertMessage += ": \(localizedDescription)"
```

alert-2

Comment (sven-straubinger) [inline Sunrise/ViewModels/BaseViewModel.swift]: Maybe we should also change the order of `localizedFailureReason` and `localizedDescription`, since the first one is optional, but the second one isn't. When the first one is `nil`, we have error messages starting with a colon.

Comment (nikola-mladenovic) [inline Sunrise/ViewModels/BaseViewModel.swift]: Actually they're both optional for errors from the SDK, but we can juggle our way around it, so we always present as much info as possible.

File: Sunrise/ViewModels/CartViewModel.swift [hunk 7]
```diff
@@ -39,31 +39,35 @@ class CartViewModel: BaseViewModel {
 
     // MARK: - Data Source
 
-    func numberOfItemsInSection(section: Int) -> Int {
+    func numberOfRowsInSection(section: Int) -> Int {
         return cart.value?.lineItems?.count ?? 0
     }
 
-    func itemNameAtIndexPath(indexPath: NSIndexPath) -> String {
+    func cartItemNameAtIndexPath(indexPath: NSIndexPath) -> String {
```
Comment (lauraluiz) [inline Sunrise/ViewModels/CartViewModel.swift]: Maybe I was unclear, but I meant `lineItemNameAtIndexPath`, not `cartItemNameAtIndexPath`. The `cart` prefix is not important since you are in the `CartViewModel`, but the `lineItem` name is important.

File: Sunrise/ViewModels/CartViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    private func calculateOrderDiscount() -> String {
+        guard let lineItems = cart.value?.lineItems else { return "" }
+
+        let totalOrderDiscountAmount = lineItems.reduce(0, combine: { $0 + calculateCartDiscountForLineItem($1) })
+        return Money(currencyCode: lineItems.first?.totalPrice?.currencyCode ?? "",
+                     centAmount: totalOrderDiscountAmount).description
+
+    }
+
+    private func calculateCartDiscountForLineItem(lineItem: LineItem) -> Int {
```
Comment (nikola-mladenovic) [inline Sunrise/ViewModels/CartViewModel.swift]: @cneijenhuis @mmoelli - this is the part where I could really use your pair of eyes to double check the discount logic..

File: Sunrise/Models/DiscountedLineItemPriceForQuantity.swift [hunk 9]
```diff
@@ -0,0 +1,23 @@
+//
+// Copyright (c) 2016 Commercetools. All rights reserved.
+//
+
+import ObjectMapper
+
+struct DiscountedLineItemPrice: Mappable {
```
Comment (sven-straubinger) [inline Sunrise/Models/DiscountedLineItemPriceForQuantity.swift]: Is it there a reason the file-name differs slightly from the struct it contains?
 `DiscountedLineItemPriceForQuantity.swift` vs. `DiscountedLineItemPrice` ?

File: Sunrise/Models/DiscountedPricePerQuantity.swift [hunk 10]
```diff
@@ -0,0 +1,23 @@
+//
+// Copyright (c) 2016 Commercetools. All rights reserved.
+//
+
+import ObjectMapper
+
+struct DiscountedLineItemPriceForQuantity: Mappable {
```
Comment (sven-straubinger) [inline Sunrise/Models/DiscountedPricePerQuantity.swift]: Same here:
 `DiscountedPricePerQuantity.swift` vs. `DiscountedLineItemPriceForQuantity` ?

File: Sunrise/Models/DiscountedLineItemPriceForQuantity.swift (hunk 9 above)
Comment (nikola-mladenovic) [inline Sunrise/Models/DiscountedLineItemPriceForQuantity.swift]: No reason, good catch 👍

File: Sunrise/ViewModels/BaseViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        self.alertMessageObserver = alertMessageObserver
+    }
+
+    func alertMessageForErrors(errors: [NSError]) -> String {
+        return errors.map({
+            var alertMessage = ""
+            if let failureReason = $0.userInfo[NSLocalizedFailureReasonErrorKey] as? String {
+                alertMessage += "\(failureReason) :"
+            }
+            alertMessage += $0.userInfo[NSLocalizedDescriptionKey] as? String ?? $0.localizedDescription
```
Comment (sven-straubinger) [inline Sunrise/ViewModels/BaseViewModel.swift]: Is there a benefit using the `userInfo` dictionary here, since `$0.localizedDescription` also uses it internally and also has a fallback, if the dictionary is empty?

https://developer.apple.com/library/ios/documentation/Cocoa/Reference/Foundation/Classes/NSError_Class/#//apple_ref/occ/instp/NSError/localizedDescription

File: Sunrise/ViewModels/CartViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        guard let discountedPricePerQuantity = lineItem.discountedPricePerQuantity, quantity = lineItem.quantity else { return 0 }
+
+        let discountedAmount = discountedPricePerQuantity.reduce(0, combine: {
```
Comment (nikola-mladenovic) [inline Sunrise/ViewModels/CartViewModel.swift]: @cneijenhuis @mmoelli - this is the part where I could really use your pair of eyes to double check the discount logic...

File: Sunrise/ViewControllers/CartViewController.swift [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        viewModel.cart.producer
+        .observeOn(UIScheduler())
+        .startWithNext { [weak self] _ in
+            self?.tableView.reloadData()
+        }
+
+        observeAlertMessageSignal(viewModel: viewModel)
+    }
+
+    private var cartSummaryCell: CartSummaryCell? {
```
Comment (sven-straubinger) [inline Sunrise/ViewControllers/CartViewController.swift]: Seems like this private property is never used.

File: Sunrise/ViewModels/CartViewModel.swift (hunk 12 above)
Comment (mmoelli) [inline Sunrise/ViewModels/CartViewModel.swift]: @nikola-mladenovic hmm, you may have to work me through that code.
